File: project/pro/python/dining_savages.py
#while creating object initialize the servings and NUM_SAVAGES

import thread
import threading
import time
import random

sem_emptypot = threading.Semaphore()
sem_fullpot = threading.Semaphore()

class Savages_cook:
	global NUM_SAVAGES
	global servings
	global meals

	def __init__(self,servings,NUM_SAVAGES):
		self.servings=servings
		self.NUM_SAVAGES=NUM_SAVAGES
		self.meals=0
		self.servings_thread=threading.Lock()
		self.print_thread=threading.Lock()


	def cook(self, cook_id):
		self.meals = 2
		while (self.meals != 0):
			sem_emptypot.acquire()

			self.putServingsInPot(15)
			self.meals=self.meals-1

			self.print_thread.acquire()
			print "Cook filled pot"
			self.print_thread.release()

			for i in range(0,self.NUM_SAVAGES):
				sem_fullpot.release()
		return 0

	def savage(self, savage_id):
		self.meals=11
		while ( self.meals != 0 ):

			self.servings_thread.acquire()

			myServing = self.getServingsFromPot()

			if (self.servings == 0):
				sem_fullpot.acquire()
				myServing = self.getServingsFromPot()

			self.servings_thread.release()

			self.meals=self.meals-1

			self.print_thread.acquire()
			print "Savage ", savage_id, "is eating "
			self.print_thread.release()

			time.sleep(2)

			self.print_thread.acquire()
			print "Savage ", savage_id, "is DONE eating"
			self.print_thread.release()

		return 0

	def getServingsFromPot(self):
		if (self.servings <= 0) :
			sem_emptypot.release()
			retVal=0
		else :
			self.servings=self.servings-1
			retVal=1
		return retVal

	def putServingsInPot(self,num):
		self.servings = self.servings - num;
		sem_fullpot.release()



def main():
	sav_id = [ ]
	SAVAGES = [ ]
	NUM_SAVAGES=3
	for i in range(NUM_SAVAGES+1):
		#print i
		sav_id.append(0)
	Sav_Co=Savages_cook(15,3)
	for i in range(0,NUM_SAVAGES):
		sav_id[i]=i+1;
		#print sav_id[i]
		SAVAGES=threading.Thread(target = Sav_Co.savage, args = (sav_id[i],))
		SAVAGES.start()

	SAVAGES=threading.Thread(target = Sav_Co.cook, args = (sav_id[i],))
	SAVAGES.start()

	for i in range(0,NUM_SAVAGES):	
		SAVAGES.join()



if __name__ == '__main__':
	main()

File: project/pro/C/dining_savages.c
#include <stdio.h>
#include <unistd.h>

#include <pthread.h>
#include <semaphore.h>

#define NUM_SAVAGES 3

sem_t emptyPot;
sem_t fullPot;

void *savage (void*);
void *cook (void*);

static pthread_mutex_t print_mutex;

static int servings = 15;

int getServingsFromPot(int savage_id)
 {

    if (servings > 0) 
    {
        servings--;
        printf("-------serving%d------\n",servings);     
        
        sleep(2);

        pthread_mutex_lock (&print_mutex);
        printf ("Savage: %i is DONE eating\n", savage_id);
        pthread_mutex_unlock (&print_mutex);   //signaling the cook
       
    } 
}

void putServingsInPot () {

    servings =15;
    sem_post (&fullPot);

}

void *cook (void *id)
 {
    int cook_id = *(int *)id;
    int i;

    while ( 1 ) 
    {

        sem_wait (&emptyPot); //waiting for signal

        putServingsInPot ();

        pthread_mutex_lock (&print_mutex);
        printf ("\nCook filled pot\n\n");
        pthread_mutex_unlock (&print_mutex);


    }
    pthread_exit(0);
    return NULL;
}

void *savage (void *id) 
{
    int savage_id = *(int *)id;
    int myServing;

  //printf("Hello Savage---> %d\n", savage_id);
    sleep(2);

    while (1)             
     {

        if (servings == 0)            
        {
            sem_post (&emptyPot);          //signaling the cook
            sem_wait (&fullPot);          //waiting for cook to fill the pot
            
        }
	  pthread_mutex_lock (&print_mutex);
        printf ("Savage: %d is eating\n", savage_id); 
        pthread_mutex_unlock (&print_mutex);          //to ensure more than one savage can eat 

	  getServingsFromPot(savage_id);
		       
        
    }
    pthread_exit(0);
    return NULL;
}

int main() {

    int i, id[NUM_SAVAGES+1];
    pthread_t tid[NUM_SAVAGES+1];

    // Initialize the mutex locks
    pthread_mutex_init(&print_mutex, NULL);

    // Initialize the semaphores
    sem_init(&emptyPot, 0, 0);
    sem_init(&fullPot,  0, 0);

    for (i=0; i<NUM_SAVAGES; i++) 
    {
        id[i] = i;
        printf("Savage %d entered \n", id[i]);
        pthread_create (&tid[i], NULL, savage, (void *)&id[i]);
    }
        
    pthread_create (&tid[i], NULL, cook, (void *)&id[i]);

    for (i=0; i<NUM_SAVAGES; i++)
        pthread_join(tid[i], NULL);

}
